File: src-tauri/src/workspace.rs
use serde::Serialize;
use std::path::{Component, Path, PathBuf};

const MAX_TREE_ENTRIES: usize = 5_000;
const DEFAULT_MAX_BYTES: u64 = 2 * 1024 * 1024;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TreeEntry {
    name: String,
    path: String,
    kind: &'static str,
    size: Option<u64>,
    children: Option<Vec<TreeEntry>>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FileContent {
    path: String,
    content: String,
    size: u64,
    truncated: bool,
}

fn validate_relative(relative: &str) -> Result<PathBuf, String> {
    if relative.contains('\0') {
        return Err("Path contains a NUL byte".into());
    }
    let path = Path::new(relative);
    if path.is_absolute()
        || path.components().any(|part| {
            matches!(
                part,
                Component::ParentDir | Component::RootDir | Component::Prefix(_)
            )
        })
    {
        return Err("Path must stay inside the project".into());
    }
    Ok(path.to_path_buf())
}

fn safe_target(root: &str, relative: &str) -> Result<(PathBuf, PathBuf), String> {
    let root =
        std::fs::canonicalize(root).map_err(|error| format!("Invalid project root: {error}"))?;
    if !root.is_dir() {
        return Err("Project root is not a directory".into());
    }
    let relative = validate_relative(relative)?;
    let target = std::fs::canonicalize(root.join(relative))
        .map_err(|error| format!("Path is unavailable: {error}"))?;
    if !target.starts_with(&root) {
        return Err("Path escapes the project root".into());
    }
    Ok((root, target))
}

fn build_tree(
    root: &Path,
    directory: &Path,
    depth: u16,
    count: &mut usize,
) -> Result<Vec<TreeEntry>, String> {
    let mut paths = std::fs::read_dir(directory)
        .map_err(|error| format!("Unable to read directory: {error}"))?
        .filter_map(Result::ok)
        .collect::<Vec<_>>();
    paths.sort_by_key(|entry| {
        let is_file = entry.file_type().map(|kind| kind.is_file()).unwrap_or(true);
        (is_file, entry.file_name().to_string_lossy().to_lowercase())
    });

    let mut result = Vec::new();
    for entry in paths {
        if *count >= MAX_TREE_ENTRIES {
            break;
        }
        let name = entry.file_name().to_string_lossy().into_owned();
        if matches!(name.as_str(), ".git" | "node_modules" | "target" | "dist") {
            continue;
        }
        let metadata =
            std::fs::symlink_metadata(entry.path()).map_err(|error| error.to_string())?;
        if metadata.file_type().is_symlink() {
            continue;
        }
        *count += 1;
        let path = entry.path();
        let relative = path
            .strip_prefix(root)
            .map_err(|error| error.to_string())?
            .to_string_lossy()
            .into_owned();
        if metadata.is_dir() {
            let children = if depth > 0 {
                Some(build_tree(root, &path, depth - 1, count)?)
            } else {
                None
            };
            result.push(TreeEntry {
                name,
                path: relative,
                kind: "directory",
                size: None,
                children,
            });
        } else if metadata.is_file() {
            result.push(TreeEntry {
                name,
                path: relative,
                kind: "file",
                size: Some(metadata.len()),
                children: None,
            });
        }
    }
    Ok(result)
}

#[tauri::command]
pub fn workspace_tree(
    root: String,
    relative: Option<String>,
    depth: Option<u16>,
) -> Result<Vec<TreeEntry>, String> {
    let relative = relative.unwrap_or_default();
    let (canonical_root, target) = safe_target(&root, &relative)?;
    if !target.is_dir() {
        return Err("Requested path is not a directory".into());
    }
    let mut count = 0;
    build_tree(
        &canonical_root,
        &target,
        depth.unwrap_or(3).min(8),
        &mut count,
    )
}

#[tauri::command]
pub fn workspace_read_file(
    root: String,
    relative: String,
    max_bytes: Option<u64>,
) -> Result<FileContent, String> {
    let (_, target) = safe_target(&root, &relative)?;
    let metadata = target.metadata().map_err(|error| error.to_string())?;
    if !metadata.is_file() {
        return Err("Requested path is not a file".into());
    }
    let limit = max_bytes.unwrap_or(DEFAULT_MAX_BYTES).min(10 * 1024 * 1024);
    let bytes = std::fs::read(&target).map_err(|error| format!("Unable to read file: {error}"))?;
    let truncated = bytes.len() as u64 > limit;
    let visible = &bytes[..bytes.len().min(limit as usize)];
    let content = String::from_utf8(visible.to_vec())
        .map_err(|_| "Binary files cannot be displayed".to_string())?;
    Ok(FileContent {
        path: relative,
        content,
        size: metadata.len(),
        truncated,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_unsafe_relative_paths() {
        assert!(validate_relative("src/main.ts").is_ok());
        assert!(validate_relative("../secret").is_err());
        assert!(validate_relative("/etc/passwd").is_err());
    }
}

File: src-server/workspace.ts
import { readdirSync, statSync, readFileSync, realpathSync } from "node:fs"
import { join, relative, sep, isAbsolute } from "node:path"

const MAX_TREE_ENTRIES = 5_000
const DEFAULT_MAX_BYTES = 2 * 1024 * 1024

export interface TreeEntry {
  name: string
  path: string
  kind: "file" | "directory"
  size?: number
  children?: TreeEntry[]
}

export interface FileContent {
  path: string
  content: string
  size: number
  truncated: boolean
}

function validateRelative(rel: string): string {
  if (rel.includes("\0")) throw new Error("Path contains a NUL byte")
  if (isAbsolute(rel)) throw new Error("Path must stay inside the project")
  const parts = rel.split(sep).filter((p) => p && p !== ".")
  for (const part of parts) {
    if (part === "..") throw new Error("Path must stay inside the project")
  }
  return rel
}

function safeTarget(root: string, rel: string): { root: string; target: string } {
  const canonicalRoot = realpathSync(root)
  if (!statSync(canonicalRoot).isDirectory())
    throw new Error("Project root is not a directory")
  const validated = validateRelative(rel)
  const target = realpathSync(join(canonicalRoot, validated))
  if (!target.startsWith(canonicalRoot))
    throw new Error("Path escapes the project root")
  return { root: canonicalRoot, target }
}

const IGNORED = new Set([".git", "node_modules", "target", "dist"])

function buildTree(
  root: string,
  directory: string,
  depth: number,
  count: { value: number },
): TreeEntry[] {
  let entries: string[]
  try {
    entries = readdirSync(directory)
  } catch (e: any) {
    throw new Error(`Unable to read directory: ${e.message ?? e}`)
  }
  entries.sort((a, b) => {
    const aStat = statSafe(join(directory, a))
    const bStat = statSafe(join(directory, b))
    const aFile = aStat ? !aStat.isDirectory() : true
    const bFile = bStat ? !bStat.isDirectory() : true
    if (aFile !== bFile) return aFile ? 1 : -1
    return a.toLowerCase().localeCompare(b.toLowerCase())
  })

  const result: TreeEntry[] = []
  for (const name of entries) {
    if (count.value >= MAX_TREE_ENTRIES) break
    if (IGNORED.has(name)) continue
    const full = join(directory, name)
    const meta = statSafe(full)
    if (!meta) continue
    if (meta.isSymbolicLink?.()) continue
    count.value += 1
    const rel = relative(root, full)
    if (meta.isDirectory()) {
      const children = depth > 0 ? buildTree(root, full, depth - 1, count) : undefined
      result.push({ name, path: rel, kind: "directory", children })
    } else if (meta.isFile()) {
      result.push({ name, path: rel, kind: "file", size: meta.size })
    }
  }
  return result
}

function statSafe(path: string) {
  try {
    return statSync(path, { throwIfNoEntry: false })
  } catch {
    return undefined
  }
}

export function workspaceTree(
  root: string,
  rel?: string,
  depth?: number,
): TreeEntry[] {
  const relative0 = rel ?? ""
  const { root: canonicalRoot, target } = safeTarget(root, relative0)
  if (!statSync(target).isDirectory())
    throw new Error("Requested path is not a directory")
  const count = { value: 0 }
  return buildTree(canonicalRoot, target, Math.min(depth ?? 3, 8), count)
}

export function workspaceReadFile(
  root: string,
  rel: string,
  maxBytes?: number,
): FileContent {
  const { target } = safeTarget(root, rel)
  const meta = statSync(target)
  if (!meta.isFile()) throw new Error("Requested path is not a file")
  const limit = Math.min(maxBytes ?? DEFAULT_MAX_BYTES, 10 * 1024 * 1024)
  const buffer = readFileSync(target)
  const truncated = buffer.length > limit
  const visible = buffer.subarray(0, Math.min(buffer.length, limit))
  let content: string
  try {
    content = new TextDecoder().decode(visible)
  } catch {
    throw new Error("Binary files cannot be displayed")
  }
  return { path: rel, content, size: meta.size, truncated }
}
